### PR TITLE
SCP-2136 Inline single occurence bounded terms

### PR DESCRIPTION
**Pull request by jakzale**
Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-core/plutus-ir/src/PlutusIR/Compiler.hs [hunk 1]
```diff
@@ -40,21 +46,50 @@ import qualified PlutusIR.Transform.Unwrap          as Unwrap
...
 
 -- | Actual simplifier
 simplify
-    :: (M.ToBuiltinMeaning uni fun, PLC.MonadQuote m)
+    :: forall m e uni fun a b. Compiling m e uni fun a
     => Term TyName Name uni fun b -> m (Term TyName Name uni fun b)
-simplify = DeadCode.removeDeadBindings <=< Inline.inline . Beta.beta . Unwrap.unwrapCancel
+simplify term =
+  do
+    selectedPasses <- map snd <$> filterM (shouldRunPass . fst) availablePasses
```
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Compiler.hs:61]: I really like `witherable` for writing more "pointful" versions of things like this, and we even have a dependency on it already. Something like:
```
selectedPasses <- forMaybe availablePasses $ \(optionAccessor, pass) -> do
    shouldRun <- shouldRunPass optionAccessor
    if shouldRun then pure $ Just pass else pure Nothing
```

Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Compiler.hs:61]: Alternatively, you already wrote `runIf`, you could also map `uncurry runIf` over the list, turning the non-running passes into no-ops!

File: plutus-tx-plugin/test/IsData/deconstructData.plc.golden [hunk 2]
```diff
@@ -1,6 +1,24 @@
 (program
   (let
     (nonrec)
+    (termbind
```
Comment (michaelpj) [inline plutus-tx-plugin/test/IsData/deconstructData.plc.golden:4]: Why has the inlining got worse in a bunch of these tests?

Comment (jakzale) [reply plutus-tx-plugin/test/IsData/deconstructData.plc.golden:4]: We disabled simplifier on `plutus-tx-plugin` tests by setting `max-simplifier-iterations=0` in `plutus-tx-plugin.cabal`.

File: plutus-core/plutus-ir/src/PlutusIR/Compiler.hs (hunk 1 above)
Comment (jakzale) [reply plutus-core/plutus-ir/src/PlutusIR/Compiler.hs:61]: @michaelpj I moved those changes to a separate PR:
https://github.com/input-output-hk/plutus/pull/3655

I was thinking to implement it more like:
https://github.com/input-output-hk/plutus/blob/46abe903143e092a0877586d901b1afe9d5a12db/plutus-core/plutus-ir/src/PlutusIR/Compiler.hs#L60-L106

**Comment by jakzale** [comment]
@michaelpj I think you will have to take this from here.

Note that general pir compiler improvements got split into a separate PR (https://github.com/input-output-hk/plutus/pull/3655) and should be safe to merge by now.  Nevertheless, as this PR depends on that work, they are included in this PR as well.

Unfortunately, there still seems to be a bug in the `floatTerm` pass:  compiling `plutus-ledger` will lead to the following output

```sh
[19 of 25] Compiling Ledger.Typed.Scripts.MonetaryPolicies ( src/Ledger/Typed/Scripts/MonetaryPolicies.hs, /home/jakub/Developer/plutus-inline/dist-newstyle/build/x86_64-linux/ghc-8.10.4.20210212/plutus-ledger-0.1.0.0/build/Ledger/Typed/Scripts/MonetaryPolicies.o, /home/jakub/Developer/plutus-inline/dist-newstyle/build/x86_64-linux/ghc-8.10.4.20210212/plutus-ledger-0.1.0.0/build/Ledger/Typed/Scripts/MonetaryPolicies.dyn_o )
  !!! rename
  !!! simplifyTerm
    !!! simplifier pass 1
      !!! unwrap cancel
      !!! beta
      !!! inline
      !!! remove dead bindings
    !!! simplifier pass 2
      !!! unwrap cancel
      !!! beta
      !!! inline
      !!! remove dead bindings
    !!! simplifier pass 3
      !!! unwrap cancel
      !!! beta
      !!! inline
      !!! remove dead bindings
    !!! simplifier pass 4
      !!! unwrap cancel
      !!! beta
      !!! inline
      !!! remove dead bindings
    !!! simplifier pass 5
      !!! unwrap cancel
      !!! beta
      !!! inline
      !!! remove dead bindings
    !!! simplifier pass 6
      !!! unwrap cancel
      !!! beta
      !!! inline
      !!! remove dead bindings
    !!! simplifier pass 7
      !!! unwrap cancel
      !!! beta
      !!! inline
      !!! remove dead bindings
    !!! simplifier pass 8
      !!! unwrap cancel
      !!! beta
      !!! inline
      !!! remove dead bindings
  !!! floatTerm
GHC Core to PLC plugin: E020:Error: Error from the PIR compiler:
            E020: Error during PIR typechecking:
            Free variable at  () :  d_25317
```


Also, `equality does not survive renaming without marking` seems to fail once in a blue moon:

```sh
cd plutus-core/
for run in {1..100}; do
  cabal run plutus-core:plutus-core-test -- --pattern "equality does not survive renaming without marking"
done
```

will eventually get it to fail.  I am not sure if that is simply because it generates a term with no names.

**Comment by michaelpj** [comment]
Okay, I'll take it from here.

**Comment by michaelpj** [comment]
On the plus side, I managed to isolate and minimize the let-floating bug, now let's see if I can fix it.

**Comment by jakzale** [comment]
I rebased this branch on top of #3823 and got pass pass floating let terms.  Now there seems to be an issue with compiling recursive terms.  (Note that this is with pedantic typechecking on):

```sh
[20 of 26] Compiling Ledger.Typed.Scripts.MonetaryPolicies ( src/Ledger/Typed/Scripts/MonetaryPolicies.hs, /home/jakub/Developer/plutus-inline/dist-newstyle/build/x86_64-linux/ghc-8.10.4.20210212/plutus-ledger-0.1.0.0/build/Ledger/Typed/Scripts/MonetaryPolicies.o, /home/jakub/Developer/plutus-inline/dist-newstyle/build/x86_64-linux/ghc-8.10.4.20210212/plutus-ledger-0.1.0.0/build/Ledger/Typed/Scripts/MonetaryPolicies.dyn_o )
  !!! rename
  !!! removeDeadBindings
  !!! simplifyTerm
    !!! simplifier pass 1
      !!! unwrap cancel
      !!! beta
      !!! inline
    !!! simplifier pass 2
      !!! unwrap cancel
      !!! beta
      !!! inline
    !!! simplifier pass 3
      !!! unwrap cancel
      !!! beta
      !!! inline
    !!! simplifier pass 4
      !!! unwrap cancel
      !!! beta
      !!! inline
    !!! simplifier pass 5
      !!! unwrap cancel
      !!! beta
      !!! inline
    !!! simplifier pass 6
      !!! unwrap cancel
      !!! beta
      !!! inline
    !!! simplifier pass 7
      !!! unwrap cancel
      !!! beta
      !!! inline
    !!! simplifier pass 8
      !!! unwrap cancel
      !!! beta
      !!! inline
  !!! floatTerm
  !!! compileNonStrictBindings
  !!! thunkRecursions
  !!! compileLets DataTypes
  !!! compileLets RecTerms
GHC Core to PLC plugin: E016:Error: Error from the PIR compiler:
            E016: Error during PIR typechecking:
            Type mismatch at <unknown> in term
              's'.
            Expected type
              '(ifix
                (lam
                  self_21843
                  (fun (type) (type))
                  (lam a_21844 (type) (fun [ self_21843 a_21843 ] a_21843))
                )
                a_0
              )',
            found type
              '(ifix
                (lam
                  self_21832
                  (fun (type) (type))
                  (lam a_21833 (type) (fun [ self_21832 a_21832 ] a_21832))
                )
                (lam
                  a_21837
                  (type)
                  (ifix
                    (lam
                      self_21838
                      (fun (type) (type))
                      (lam a_21839 (type) (fun [ self_21838 a_21838 ] a_21838))
                    )
                    a_21837
                  )
                )
              )'
Context: Compiling expr at "plutus-ledger-0.1.0.0-inplace:Ledger.Typed.Scripts.MonetaryPolicies:(53,11)-(53,105)"
```

**Comment by jakzale** [comment]
Suspected there was something off with type inlining being disabled, but after enabling it, it made stuff worse.

Added it as a reverted commit to `jakzale/inline` for posterity.

**Pull request closed without merging**